### PR TITLE
ADBDEV-5551: Fix gprestore handle copy error with jobs > 1

### PR DESCRIPTION
**Pull request by RekGRpth**
Fix gprestore handle copy error with jobs > 1.

When using the jobs > 1 option, if a copy error occurred, the corresponding
goroutine was terminated. But other goroutines continued to wait for COPY and
restore froze.

This patch adds the use of execution context to goroutines, which allows
requests to be canceled on errors.

### CODE REVIEW

**Comment by Stolb27** [comment]
> The test was added only for restore, because with backup it is quite difficult
to reproduce the error when copying a table.

The restore part LGTM. Separate out backup code part to the separate task with lower priority until you reproduce it and update branch.

**Comment by RekGRpth** [comment]
> > The test was added only for restore, because with backup it is quite difficult
> > to reproduce the error when copying a table.
> 
> The restore part LGTM. Separate out backup code part to the separate task with lower priority until you reproduce it and update branch.

done https://github.com/arenadata/gpbackup/pull/80

File: restore/data.go [hunk 1]
```diff
@@ -227,8 +230,15 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 
 			setGUCsForConnection(gucStatements, whichConn)
 			for entry := range tasks {
+				// Check if any error occurred in any other goroutines:
+				select {
+				case <-ctx.Done():
+					return // Error somewhere, terminate
```
Comment (red1452) [inline restore/data.go:236]: Why did not you add `cancel()` here?

Comment (RekGRpth) [reply restore/data.go:236]: because cancels in other places lead to this case of select

File: end_to_end/end_to_end_suite_test.go [hunk 2]
```diff
@@ -2263,6 +2263,21 @@ LANGUAGE plpgsql NO SQL;`)
 			assertArtifactsCleaned(restoreConn, "20240502095933")
 			testhelper.AssertQueryRuns(restoreConn, "DROP TABLE t0; DROP TABLE t1; DROP TABLE t2; DROP TABLE t3; DROP TABLE t4;")
 		})
+		It("Will not hang after error during restore with jobs", func() {
+			command := exec.Command("tar", "-xzf", "resources/2-segment-db-error.tar.gz", "-C", backupDir)
```
Comment (Stolb27) [inline end_to_end/end_to_end_suite_test.go:2267]: This test can't be run on demo cluster:

```
      20240506:14:39:43 gprestore:givi:samsung-ivanov:120268-[INFO]:-Restoring pre-data metadata
Pre-data objects restored:  0 / 14 [-------------------------------------------------------]   0.00%20240506:14:39:43 gprestore:givi:samsung-ivanov:120268-[CRITICAL]:-ERROR: role "gpadmin" does not exist (SQLSTATE 42704): Error encountered while creating schema public
```

Comment (RekGRpth) [reply end_to_end/end_to_end_suite_test.go:2267]: I run this test on demo cluster, moreover data for this test is taken from another merged patch

Comment (Stolb27) [reply end_to_end/end_to_end_suite_test.go:2267]: this test should be reworked later

**Review by red1452 — APPROVED** [review]

**Pull request merged**